### PR TITLE
[test] feature: Add utility to generate coverage report based on test function metadata

### PR DESCRIPTION
**Pull request by arjkesh**
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]

*Description:*
- Add utilities to pytest in order to generate reports based on markers, fixtures and other metadata
- Update instance types to be multi-gpu instances when sensible

*Tests run:*
- Executed command below, generated csv files
- Added a test to ensure the command below runs without error
- Added a test to ensure that we are not running on single gpu instance types unless we've explicitly allowed running on those instance types

Example Failure:
```
...
E           INTERNALERROR> 1. Function <local_path>deep-learning-containers/test/dlc_tests/ecs/pytorch/inference/test_ecs_pytorch_inference.py::test_ecs_pytorch_inference_gpu uses single-gpu instance type p3.2xlarge. Please use multi-gpu instance type.
E           INTERNALERROR> ******Problems with <local_path>deep-learning-containers/test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py::test_ecs_pytorch_training_dgl_gpu:******
E           INTERNALERROR> 1. Function <local_path>deep-learning-containers/test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py::test_ecs_pytorch_training_dgl_gpu uses single-gpu instance type p2.xlarge. Please use multi-gpu instance type.
E           INTERNALERROR> ******Problems with <local_path>deep-learning-containers/test/dlc_tests/ecs/tensorflow/training/test_ecs_tensorflow_training.py::test_ecs_tensorflow_training_mnist_gpu:******
E           INTERNALERROR> 1. Function <local_path>deep-learning-containers/test/dlc_tests/ecs/tensorflow/training/test_ecs_tensorflow_training.py::test_ecs_tensorflow_training_mnist_gpu uses single-gpu instance type p3.2xlarge. Please use multi-gpu instance type.
E           INTERNALERROR> ******Problems with <local_path>deep-learning-containers/test/dlc_tests/sanity/test_pre_release.py::test_framework_version_gpu:******
E           INTERNALERROR> 1. Function <local_path>deep-learning-containers/test/dlc_tests/sanity/test_pre_release.py::test_framework_version_gpu uses single-gpu instance type p2.xlarge. Please use multi-gpu instance type.
E           INTERNALERROR> TOTAL ISSUES: 30
```

*Additional context:*
Usage (root dir is dlc_tests)
- ensure you are logged into an aws account with `aws sts get-caller-identity` -- this is necessary to use boto3 to determine the number of gpus in a gpu instance type
```
export DLC_IMAGES='test' && pytest -s sanity/test_generate_coverage_doc.py
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/conftest.py [hunk 1]
```diff
@@ -222,10 +235,105 @@ def pytest_runtest_setup(item):
...
+            function_name = item.name.split("[")[0]
+            function_key = f"{item.fspath}::{function_name}"
+            str_fspath = str(item.fspath)
+            str_keywords = str(item.keywords)
+
+            # Based on keywords and filepaths, assign values
+            scope = infer_field_value("all", ("mxnet", "tensorflow", "pytorch"), str_fspath)
+            train_inf = infer_field_value("both", ("training", "inference"), str_fspath, str_keywords)
+            integration = infer_field_value("general integration", ("_dgl_", "_smdebug_", "_gluonnlp_"), str_keywords)
+            model = infer_field_value("N/A", ("mnist", "densenet"), str_keywords)
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:252]: These are basically to guess what the integration is based on the test name and other metadata in the case where someone did not use the integration marker.

File: test/dlc_tests/conftest.py [hunk 2]
```diff
@@ -213,6 +217,15 @@ def pytest_configure(config):
     config.addinivalue_line(
         "markers", "canary(message): mark test to run as a part of canary tests."
     )
+    config.addinivalue_line(
+        "markers", "integration(ml_integration): mark what the test is testing."
+    )
+    config.addinivalue_line(
+        "markers", "model(model_name): The name of the pytest model being tested"
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:224]: nit: pytest model is incorrect here, should be ML model

File: test/dlc_tests/conftest.py [hunk 3]
```diff
@@ -213,6 +217,15 @@ def pytest_configure(config):
...
+    )
+    config.addinivalue_line(
+        "markers", "multinode(num_instances): The number of instances the test is run on"
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:227]: This is an optional marker. If it is not used, we assume single node

Comment (arjkesh) [reply test/dlc_tests/conftest.py:227]: no longer relevant -- we will write "multinode" if we find multinode in the function metadata

File: test/dlc_tests/conftest.py [hunk 4]
```diff
@@ -222,10 +235,111 @@ def pytest_runtest_setup(item):
...
+
+
+def get_marker_arg_value(item_obj, marker_name, default=None):
+    """
+    Function to return the argument value of a pytest marker -- if it does not exist, fall back to a default.
+    If the default does not exist and the option does not exist, raise an error.
+
+    :param item_obj: pytest item object
+    :param marker_name: name of the pytest marker
+    :param default: default return value -- if none,
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:326]: nit: fill in the rest of this docstring. If None, assume this is a required marker

File: test/dlc_tests/sanity/test_generate_coverage_doc.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+@pytest.mark.integration("Generating this coverage doc")
+def test_generate_coverage_doc():
+    """
+    Test generating the test coverage doc
+    """
+    test_coverage_file = TEST_COVERAGE_FILE
+    ctx = Context()
+    # Set DLC_TESTS to 'test' to avoid image names affecting function metadata (due to parametrization)
+    # Set CODEBUILD_RESOLVED_SOURCE_VERSION to test for ease of running this test locally
+    ctx.run("export DLC_TESTS='test' && export CODEBUILD_RESOLVED_SOURCE_VERSION='test' "
```
Comment (arjkesh) [inline test/dlc_tests/sanity/test_generate_coverage_doc.py:18]: nit: I can remove CODEBUILD_RESOLVED_SOURCE_VERSION here -- mostly did this to facilitate local testing, but we have a requirement to set this env variable set in our docs anyway

File: test/dlc_tests/conftest.py [hunk 6]
```diff
@@ -222,10 +235,112 @@ def pytest_runtest_setup(item):
...
+                                            ("_dgl_", "smdebug", "gluonnlp", "smexperiments", "_mme_", "pipemode",
+                                             "tensorboard", "_s3_"),
+                                            str_keywords)
+            model = infer_field_value("N/A",
+                                      ("mnist", "densenet", "squeezenet", "half_plus_two", "half_plus_three"),
+                                      str_keywords)
+            num_instances = infer_field_value(1, ("_multinode_", "_multi-node_"), str_fspath, str_keywords)
+            cpu_gpu = infer_field_value("all", ("cpu", "gpu", "eia"), str_keywords)
+            if cpu_gpu == "gpu":
+                if "p2.xlarge" in str_keywords:
```
Comment (nskool) [inline test/dlc_tests/conftest.py:262]: p3.2xlarge is also single-GPU. Could you check other instance types [g*] that might be single-GPU?

File: test/dlc_tests/conftest.py [hunk 7]
```diff
@@ -222,10 +235,112 @@ def pytest_runtest_setup(item):
...
+                                            "Name": function_name,
+                                            "Scope": scope,
+                                            "Job_Type": train_inf,
+                                            "Num_Instances": get_marker_arg_value(item, "multinode", num_instances),
+                                            "Processor": cpu_gpu,
+                                            "Integration": get_marker_arg_value(item, "integration", integration),
+                                            "Model": get_marker_arg_value(item, "model", model),
+                                            "Github_Link": github_link,
+                                           }
+        with open(test_coverage_file, "w+") as tc_file:
```
Comment (nskool) [inline test/dlc_tests/conftest.py:284]: Let's separate this out into it's own function.

File: test/dlc_tests/conftest.py (hunk 6 above)
Comment (arjkesh) [reply test/dlc_tests/conftest.py:262]: I will do this via boto3 apis

File: test/dlc_tests/conftest.py (hunk 7 above)
Comment (arjkesh) [reply test/dlc_tests/conftest.py:284]: sure, good call

File: test/dlc_tests/conftest.py [hunk 8]
```diff
@@ -222,10 +236,201 @@ def pytest_runtest_setup(item):
...
+                LOGGER.warning(f"Found failure message, but no issues. Message:\n{message}")
+            else:
+                raise TestReportGenerationFailure(message)
+
+
+def _handle_single_gpu_instances(function_key, function_keywords, failures, cpu_gpu="gpu"):
+    """
+    Generally, we do not want tests running on single gpu instance types. However, there are exceptions to this rule.
+    This function is used to determine whether we need to raise an error with report generation or not, based on
+    whether we are using single gpu instanecs or not in a given test function.
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:303]: nit: typo on instances

File: test/dlc_tests/conftest.py [hunk 9]
```diff
@@ -222,10 +236,201 @@ def pytest_runtest_setup(item):
...
+                                             ("_dgl_", "smdebug", "gluonnlp", "smexperiments", "_mme_", "pipemode",
+                                             "tensorboard", "_s3_"),
+                                             str_keywords)
+            model = _infer_field_value("N/A",
+                                       ("mnist", "densenet", "squeezenet", "half_plus_two", "half_plus_three"),
+                                       str_keywords)
+            num_instances = _infer_field_value(1, ("_multinode_", "_multi-node_"), str_fspath, str_keywords)
+            cpu_gpu = _infer_field_value("all", ("cpu", "gpu", "eia"), str_keywords)
+            if cpu_gpu == "gpu":
+                cpu_gpu = _handle_single_gpu_instances(function_key, str_keywords, failure_conditions)
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:274]: update function call to be - `cpu_gpu, failure_conditions = _handle_single_gpu_instances(function_key, str_keywords, failure_conditions)`

File: test/dlc_tests/conftest.py [hunk 10]
```diff
@@ -222,10 +235,112 @@ def pytest_runtest_setup(item):
...
+        test_cov = {}
+        for item in items:
+            # Define additional csv options
+            function_name = item.name.split("[")[0]
+            function_key = f"{item.fspath}::{function_name}"
+            str_fspath = str(item.fspath)
+            str_keywords = str(item.keywords)
+
+            # Based on keywords and filepaths, assign values
+            scope = infer_field_value("all", ("mxnet", "tensorflow", "pytorch"), str_fspath)
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/dlc_tests/conftest.py]: updated, thanks

File: test/dlc_tests/conftest.py [hunk 11]
```diff
@@ -222,10 +236,203 @@ def pytest_runtest_setup(item):
...
+            integration_scope = _infer_field_value("general integration",
+                                                   ("_dgl_", "smdebug", "gluonnlp", "smexperiments", "_mme_",
+                                                    "pipemode", "tensorboard", "_s3_"),
+                                                   str_keywords)
+            model_scope = _infer_field_value("N/A",
+                                             ("mnist", "densenet", "squeezenet", "half_plus_two", "half_plus_three"),
+                                             str_keywords)
+            num_instances = _infer_field_value(1, ("_multinode_", "_multi-node_", "_multi_node_"), str_fspath,
+                                               str_keywords)
+            cpu_gpu = _infer_field_value("all", ("cpu", "gpu", "eia"), str_keywords)
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:273]: nit: change to processor_scope

File: test/dlc_tests/conftest.py [hunk 12]
```diff
@@ -113,9 +117,18 @@ def ec2_instance(
         "MaxCount": 1,
         "MinCount": 1,
     }
-    extra_volume_size_mapping = [{"DeviceName": "/dev/sda1", "Ebs": {"VolumeSize": 300, }}]
-    if ("benchmark" in os.getenv("TEST_TYPE") and (("mxnet_training" in request.fixturenames and "gpu_only" in request.fixturenames) or "mxnet_inference" in request.fixturenames)) \
-            or ("tensorflow_training" in request.fixturenames and "gpu_only" in request.fixturenames and "horovod" in ec2_key_name):
+    extra_volume_size_mapping = [{"DeviceName": "/dev/sda1", "Ebs": {"VolumeSize": 300,}}]
```
Comment (arjkesh) [inline test/dlc_tests/conftest.py:120]: this happened from running black on this file -- sort of exposes a very long and complicated if statement. it's probably worth it to factor out a helper function

File: test/dlc_tests/conftest.py [hunk 13]
```diff
@@ -222,10 +236,210 @@ def pytest_runtest_setup(item):
...
+    :param processor: whether the test is for cpu, gpu or both
+    :return: processor if not single gpu instance, else "single_gpu", and a dict with updated failure messages
+    """
+
+    # Define conditions where we allow a test function to run with a single gpu instance
+    whitelist_single_gpu = False
+    allowed_single_gpu = ("telemetry", "test_framework_version_gpu")
+
+    # Regex in order to determine the gpu instance type
+    gpu_instance_pattern = re.compile(r"\w+\.\d*xlarge")
```
Comment (nskool) [inline test/dlc_tests/conftest.py:324]: If this pattern were to be applied to ml instances, it won't work.

File: test/dlc_tests/conftest.py [hunk 14]
```diff
@@ -222,10 +236,210 @@ def pytest_runtest_setup(item):
...
+    :param function_key: local/path/to/function::function_name
+    :param function_keywords: string of keywords associated with the test function
+    :param failures: running dictionary of failures associated with the github link
+    :param processor: whether the test is for cpu, gpu or both
+    :return: processor if not single gpu instance, else "single_gpu", and a dict with updated failure messages
+    """
+
+    # Define conditions where we allow a test function to run with a single gpu instance
+    whitelist_single_gpu = False
+    allowed_single_gpu = ("telemetry", "test_framework_version_gpu")
```
Comment (nskool) [inline test/dlc_tests/conftest.py:321]: Should this list be placed in a more accessible global location? IMO that'll help making the change quickly for someone who creates a PR.

File: test/dlc_tests/conftest.py [hunk 15]
```diff
@@ -222,10 +236,210 @@ def pytest_runtest_setup(item):
...
+    Function to assemble the failure message if there are any to raise
+
+    :param failure_messages: dict where key is the function, and value is a list of failures associated with the
+    function
+    :return: the final failure message string
+    """
+    final_message = ""
+    total_issues = 0
+    for func, messages in failure_messages.items():
+        final_message += f"******Problems with {func}:******\n"
```
Comment (nskool) [inline test/dlc_tests/conftest.py:361]: This line seems redundant based on the error message shared in the description.

File: test/dlc_tests/conftest.py (hunk 13 above)
Comment (arjkesh) [reply test/dlc_tests/conftest.py:324]: This current implementation is for dlc_tests, which excludes SM. A followup PR will address report generation with SM

File: test/dlc_tests/conftest.py (hunk 15 above)
Comment (arjkesh) [reply test/dlc_tests/conftest.py:361]: The idea here is that one can add errors to a given function via failure_messages. So if there were errors to add outside of just single gpu, they would be displayed under the function that they affect

File: test/dlc_tests/conftest.py (hunk 14 above)
Comment (arjkesh) [reply test/dlc_tests/conftest.py:321]: sure, I can create a global for this in conftest

File: test/dlc_tests/conftest.py (hunk 15 above)
Comment (nskool) [reply test/dlc_tests/conftest.py:361]: Correct - the error message for multi_gpu also includes {function_key} - that can be removed as the function name is printed by this function anyway.

Comment (arjkesh) [reply test/dlc_tests/conftest.py:361]: oh i see, sure i can remove it

**Pull request merged**